### PR TITLE
[IMPROVE] UI design for Tables and tabs component on Directory

### PR DESCRIPTION
**Pull request by karlprieb**
A better design for tables and tab component, applied on Directory page.

Still needs to write a documentation about these new components and how the directory feature works. We can make a blogpost. What you @RocketChat/core think?

### CODE REVIEW

**Comment by gdelavald** [comment]
There is this weird delay when toggling between users and channels, could this be a performance issue?

**Comment by graywolf336** [comment]
Why is the name of this pull request labeled as `[FIX]` when the description says:
> New table and tab component applied on Directory page.

**Comment by ggazzo** [comment]
the design is new, but we didnt change nothing as feature, just fixed the bad UI, or am I wrong?

**Comment by geekgonecrazy** [comment]
Should we instead use an improve tag?  Instead of fix or new.

**Comment by vynmera** [comment]
@geekgonecrazy Good call, an IMPROVE tag would make the most sense here.

Furthermore, by using a tag like that, we might get more people to polish up the existing parts of RC, something quite important to good UX and code :-)

**Comment by ggazzo** [comment]
I agree @geekgonecrazy , but I thought we had just NEW, BREAK, FIX

**Comment by vynmera** [comment]
@ggazzo that's the issue, I think~

**Comment by geekgonecrazy** [comment]
@ggazzo right we only have the 3.  I'm suggesting we consider one more.  That way things such as component redesign can be surfaced separately so we can make it more clear the components we improve vs the ones we fix a bug only.

Like some times we completely rework something for the purpose of improving a lot of things and it does close bugs, but over all we improved functionality. 

Just an idea.  @rodrigok what do you think?  That or should we add a tag to PR's that it's design, that way they can be found that way?

File: packages/rocketchat-theme/client/imports/components/tabs.css [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+	width: 100%;
+
+	margin: 0 -1rem 0.5rem;
+	padding: 0 0.5rem;
+
+	border-bottom: 1px solid rgba(216,216,216, 0.2);
+}
+
+.tab {
+	margin: 0 1rem;
```
Comment (ggazzo) [inline packages/rocketchat-theme/client/imports/components/tabs.css]: I think you should use `transition: color .3s;` here

File: packages/rocketchat-ui/client/components/table.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+Template.table.onRendered(function() {
+	const dummyTr = '<tr class="table-tr-dummy"></tr>';
+	this.$('tbody').prepend(dummyTr).append(dummyTr);
+});
+
+Template.table.events({
+	'click tbody tr'() {
+		const onItemClick = Template.instance().data.onItemClick;
+		if (onItemClick) {
```
Comment (ggazzo) [inline packages/rocketchat-ui/client/components/table.js]:
```javascript
return onItemClick && onItemClick(this)
```

File: packages/rocketchat-ui/client/components/tabs.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+Template.tabs.onRendered(function() {
+	this.$('.tab').first().addClass('active');
+	this.data.tabs.onChange(this.data.tabs.tabs[0].value);
+});
+
+Template.tabs.events({
+	'click .tab'(e) {
+		$('.tab').removeClass('active');
```
Comment (ggazzo) [inline packages/rocketchat-ui/client/components/tabs.js]: I think this way will not work for nested tabs

File: packages/rocketchat-ui/client/views/app/directory.js [hunk 4]
```diff
@@ -52,6 +52,49 @@ Template.directory.helpers({
...
+			onChange(value) {
+				end.set(false);
+				sortDirection.set('asc');
+				page.set(0);
+				searchType.set(value);
+			}
+		};
+	},
+	onTableItemClick() {
+		const searchType = Template.instance().searchType;
```
Comment (ggazzo) [inline packages/rocketchat-ui/client/views/app/directory.js]: `const { searchType } = Template.instance();`

File: packages/rocketchat-ui/client/components/tabs.js (hunk 3 above)
Comment (karlprieb) [reply packages/rocketchat-ui/client/components/tabs.js]: how nested tabs should work? We will have nested tabs? We already have a design for that?

File: packages/rocketchat-theme/client/imports/components/tabs.css (hunk 1 above)
Comment (karlprieb) [reply packages/rocketchat-theme/client/imports/components/tabs.css]: 👍

File: packages/rocketchat-ui/client/components/table.js (hunk 2 above)
Comment (karlprieb) [reply packages/rocketchat-ui/client/components/table.js]: 👍

File: packages/rocketchat-ui/client/views/app/directory.js (hunk 4 above)
Comment (karlprieb) [reply packages/rocketchat-ui/client/views/app/directory.js]: 👍

File: packages/rocketchat-ui/client/components/tabs.js (hunk 3 above)
Comment (ggazzo) [reply packages/rocketchat-ui/client/components/tabs.js]: we dont have design to show you now, but just to think...not for now, but is a small change

Comment (gdelavald) [reply packages/rocketchat-ui/client/components/tabs.js]: Can we do a simple `{activeClass}` variable on the helper and change according to state?

**Comment by ggazzo** [comment]
image

**Comment by karlprieb** [comment]
All done :)

@ggazzo you already did a working infinite scroll, now I just added the loading icon while your method is getting new data and I added a debounce on the search input :)

Mobile view, new sort icons:
localhost_3000_directory iphone 6_7_8

Loading state:

File: packages/rocketchat-ui/client/components/tabs.js (hunk 3 above)
Comment (karlprieb) [reply packages/rocketchat-ui/client/components/tabs.js]: I created a 'ReactiveVar' to keep the state of the active tab and adde a helper to check which tab is active and add an active class to this tab.

Thank you for your suggestions :)

**Comment by ggazzo** [comment]
@karlprieb sure I did the infinite scroll but is not working after your changes

look here https://rocket-chat-pr-11026.herokuapp.com/directory I created enough rooms to "scroll"

**Comment by karlprieb** [comment]
Fixed :)

**Comment by gdelavald** [comment]
@karlprieb @ggazzo 
There are still some weird issues going on with the list data:
image

It's reproducible by adding a search term and then switching tabs, the same happens when switching from Users to Channels:
image

**Comment by karlprieb** [comment]
@gdelavald I didn't see that :) I'll make a fix!

**Comment by karlprieb** [comment]
@gdelavald it is fixed now 👍

**Comment by kaiiiiiiiii** [comment]
Hey @karlprieb,

two more things.

1. Everything looks a bit displaced when a channel description is given.
2. Is there a reason why the description width is still limited to 200px? Doesn't make sense on large screen IMO. Sometimes our project channels have really long descriptions (100-200 words) and it would be nice to see more than just the first 5 words :)

screen shot 2018-06-14 at 21 36 34

**Comment by karlprieb** [comment]
@kaiiiiiiiii Thank you for your report :)

I already fixed the first problem.
For the second and third problems we will need some help from our design team, so I won't be able to fix that today.

@ggazzo ☝️

**Comment by karlprieb** [comment]
Hey @kaiiiiiiiii we will remove channel description from this table layout for now. In a near future we will add a extended (with descriptions) and compact layout for this table, just like we did with the sidebar.

Thank again for your reports :)

**Pull request merged**